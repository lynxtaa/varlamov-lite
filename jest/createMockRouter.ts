import { useRouter } from 'next/navigation'
import { vi } from 'vitest'

type AppRouterInstance = ReturnType<typeof useRouter>

export function createMockRouter(
	options?: Partial<AppRouterInstance>,
): AppRouterInstance {
	return {
		push: vi.fn(),
		replace: vi.fn(),
		refresh: vi.fn(),
		prefetch: vi.fn(),
		back: vi.fn(),
		forward: vi.fn(),
		...options,
	}
}
