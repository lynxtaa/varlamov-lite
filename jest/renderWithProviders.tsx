import { render as _render, type RenderOptions } from '@testing-library/react'
import React from 'react'

export const renderWithProviders = (el: React.ReactElement, options?: RenderOptions) =>
	_render(el, { ...options })
