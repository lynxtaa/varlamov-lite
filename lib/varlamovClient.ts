import 'server-only'

import * as cheerio from 'cheerio'
import pick from 'lodash/pick'
import truncate from 'lodash/truncate'
import PQueue from 'p-queue'
import probeImageSize from 'probe-image-size'

import { createFetch, FetchFunction } from './createFetch'
import { getYoutubeVideoId } from './getYoutubeVideoId'
import { articleSchema, articlesSchema } from './schemas'

const BLOG_ID = 500000
const PAGE_SIZE = 7

export interface Article {
	id: number
	uri: string
	title: string
	published_at: string | null
}

export interface ArticleFull extends Article {
	previewImageUrl: string | null
	excerpt: string
	readingTime: number
	text: string
	tags: string[]
	topics: { name: string }[]
}

type CacheOptions = Pick<RequestInit, 'cache' | 'next'>

class VarlamovClient {
	pageSize: number
	queue: PQueue
	supportTagsWithAttrs: Record<string, string[]>
	#fetchTeletype: FetchFunction

	constructor({ pageSize = 19 }: { pageSize?: number } = {}) {
		this.pageSize = pageSize
		this.queue = new PQueue({ concurrency: 5 })

		this.supportTagsWithAttrs = {
			'a[href]': ['href'],
			'b': [],
			'blockquote': [],
			'br': [],
			'em': [],
			'figcaption': [],
			'figure': [],
			'i': [],
			'iframe': ['src', 'frameBorder', 'allow', 'allowFullScreen'],
			'img': ['src', 'width', 'height', 'alt'],
			'li': [],
			'p': [],
			'span': [],
			'strong': [],
			'ul': [],
			'section': [],
		}

		this.#fetchTeletype = createFetch({
			timeout: 20_000,
			headers: { accept: 'application/json' },
			prefixUrl: 'https://teletype.in/',
		})
	}

	private async getImageSize(url: string) {
		return this.queue.add(async () =>
			probeImageSize(encodeURI(url)).catch(err => {
				if (process.env.NODE_ENV === 'development') {
					console.warn(err)
				}
				return null
			}),
		)
	}

	async getArticles({
		lastArticle,
		limit = PAGE_SIZE,
		...options
	}: {
		lastArticle?: number
		limit?: number
	} & CacheOptions = {}): Promise<Article[]> {
		const searchParams = new URLSearchParams({ limit: String(limit) })
		if (lastArticle !== undefined) {
			searchParams.set('last_article', String(lastArticle))
		}

		const { articles } = await this.#fetchTeletype(
			`api/blogs/id/${BLOG_ID}/articles?${searchParams.toString()}`,
			{ ...options, schema: articlesSchema },
		)

		return articles
	}

	async searchArticles(
		query: string,
		{
			limit = PAGE_SIZE,
			pageNum = 1,
			...options
		}: { limit?: number; pageNum?: number } & CacheOptions = {},
	): Promise<Article[]> {
		const { articles } = await this.#fetchTeletype('api/search', {
			method: 'POST',
			json: {
				query,
				blogs: false,
				articles: true,
				blog_id: BLOG_ID,
				limit,
				offset: limit * (pageNum - 1),
			},
			...options,
			schema: articlesSchema,
		})

		return articles
	}

	async getArticle(uri: string, options?: CacheOptions): Promise<ArticleFull> {
		const article = await this.#fetchTeletype(
			`api/blogs/domain/varlamov.ru/articles/${encodeURIComponent(uri)}`,
			{ ...options, schema: articleSchema },
		)

		const $ = cheerio.load(article.text, { xmlMode: true, decodeEntities: false })

		const textEl = $('document')

		const tags = $('tag')
			.toArray()
			.map(el => $(el).text().trim())

		// Заменим конструкции <image src="http://example.com"><caption>Подпись</caption></image>
		// на <img src="http://example.com" /><caption>Подпись</caption>
		for (const image of textEl.find('image').toArray()) {
			const $image = $(image)

			$image.after('<img />')
			const $img = $image.next()
			for (const [attr, value] of Object.entries(image.attribs)) {
				$img.attr(attr, value)
			}
			const caption = $image.find('caption').html()

			if (caption !== null && caption !== '') {
				$img.wrap('<figure></figure>')
				$img.after(`<figcaption>${caption}</figcaption>`)
			}

			$image.remove()
		}

		// Заменим конструкции <youtube src="https://www.youtube.com/embed/fw5hbjqBiSA" />
		// на iframe'ы
		for (const youtube of textEl.find('youtube').toArray()) {
			const $youtube = $(youtube)
			const src = $youtube.attr('src')

			if (src !== undefined && src !== '') {
				const vidId = getYoutubeVideoId(src)

				if (vidId !== null) {
					$youtube.after('<iframe> </iframe>')
					const frame = $youtube.next()
					frame.attr('src', `https://www.youtube.com/embed/${vidId}`)
					frame.attr('frameBorder', '0')
					frame.attr(
						'allow',
						'accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture',
					)
					frame.attr('allowFullScreen', '')
				}
			}
			$youtube.remove()
		}

		textEl.find(`*:not(${Object.keys(this.supportTagsWithAttrs).join(', ')})`).remove()

		for (const [selector, attrList] of Object.entries(this.supportTagsWithAttrs)) {
			const foundElements = textEl.find(selector).toArray()
			for (const element of foundElements) {
				element.attribs = pick(element.attribs, attrList)
			}
		}

		const images = textEl.find('img').toArray()

		const [firstImage] = images

		const previewImageUrl =
			article.sharing_image !== undefined
				? article.sharing_image
				: firstImage
					? $(firstImage).attr('src')
					: undefined

		await Promise.all(
			images.map(async image => {
				const img = $(image)
				const src = img.attr('src')

				if (src === undefined || src === '') {
					img.remove()
					return
				}

				const width = img.attr('width')
				const height = img.attr('height')

				if (
					width === undefined ||
					width === '' ||
					height === undefined ||
					height === ''
				) {
					const dimensions = await this.getImageSize(src)

					if (dimensions) {
						// eslint-disable-next-line require-atomic-updates
						image.attribs = {}
						img.attr('src', src)
						img.attr('width', String(dimensions.width))
						img.attr('height', String(dimensions.height))
					} else {
						img.remove()
					}
				}
			}),
		)

		const plainText = textEl.text().trim()

		const html = textEl.html()?.trim() ?? ''

		const words = plainText.match(/\S+/g)
		const WORDS_PER_MINUTE = 200

		const excerpt = article.sharing_text ?? truncate(plainText, { length: 150 })

		const readingTime = words
			? Math.round((words.length / WORDS_PER_MINUTE) * 60 * 1000)
			: 0

		return {
			id: article.id,
			uri: article.uri,
			excerpt,
			previewImageUrl:
				previewImageUrl !== undefined && previewImageUrl !== '' ? previewImageUrl : null,
			title: article.title,
			tags,
			text: html,
			published_at: article.published_at ?? null,
			readingTime,
			topics: article.topics ?? [],
		}
	}
}

export const varlamovClient = new VarlamovClient()
