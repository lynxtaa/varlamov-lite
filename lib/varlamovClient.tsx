import { URL, URLSearchParams } from 'url'

import cheerio from 'cheerio'
import { parse as parseDate, isValid } from 'date-fns'
import { zonedTimeToUtc } from 'date-fns-tz'
import { ru } from 'date-fns/locale'
import pick from 'lodash/pick'
import truncate from 'lodash/truncate'
import fetch from 'node-fetch'
import PQueue from 'p-queue'
import probeImageSize from 'probe-image-size'

type Cheerio = ReturnType<typeof cheerio>

export interface Article {
	id: number
	title: string
	createdAt: string | null
}

export interface ArticleFull extends Article {
	previewImageUrl: string | null
	excerpt: string
	text: string
	tags: string[]
}

class VarlamovClient {
	endpoint: URL
	pageSize: number
	queue: PQueue
	supportTagsWithAttrs: Record<string, string[]>

	constructor({ endpoint, pageSize = 19 }: { endpoint: string; pageSize?: number }) {
		this.endpoint = new URL(endpoint)
		this.pageSize = pageSize
		this.queue = new PQueue({ concurrency: 5 })

		this.supportTagsWithAttrs = {
			br: [],
			'a[href]': ['href'],
			p: [],
			img: ['src', 'width', 'height', 'alt'],
			i: [],
			b: [],
			em: [],
			strong: [],
			span: [],
		}
	}

	private async getHtml(path: string) {
		const response = await fetch(`${this.endpoint.toString()}${path}`)

		if (!response.ok) {
			throw new Error(`Error loading ${response.url}: ${response.status}`)
		}

		const html = await response.text()

		return html
	}

	private getImageSize(url: string) {
		return this.queue.add(() =>
			probeImageSize(encodeURI(url)).catch(err => {
				if (process.env.NODE_ENV === 'development') {
					console.warn(err)
				}
				return null
			}),
		)
	}

	private parseDate(el: Cheerio): Date | null {
		const text = el.text().trim()

		for (const format of ['d MMMM yyyy, HH:mm', 'd MMMM yyyy']) {
			const date = parseDate(text, format, new Date(), { locale: ru })

			if (isValid(date)) {
				return zonedTimeToUtc(date, 'Europe/Moscow')
			}
		}

		return null
	}

	async getArticles({
		pageNum,
		tag,
	}: {
		pageNum: number
		tag?: string
	}): Promise<Article[]> {
		const qs = new URLSearchParams()

		if (tag) {
			qs.append('tag', tag)
		}
		if (pageNum > 1) {
			qs.append('skip', String((pageNum - 1) * this.pageSize))
		}

		const html = await this.getHtml(`/?${qs.toString()}`)

		const $ = cheerio.load(html)

		const articles = $('article[id^="entry-varlamov.ru"]')
			.toArray()
			.flatMap(el => {
				const titleLink = $('.j-e-title a', el)

				const match = /(?<id>\d+)\.html$/.exec(
					// https://varlamov.ru/4130750.html
					titleLink.attr('href') || '',
				)

				if (!match?.groups) {
					return []
				}

				const createdAt = this.parseDate($('time[itemprop="dateCreated"]', el))

				return [
					{
						id: Number(match.groups.id),
						createdAt: createdAt ? createdAt.toISOString() : null,
						title: titleLink.text(),
					},
				]
			})

		return articles
	}

	async getArticle(id: number): Promise<ArticleFull> {
		let html = await this.getHtml(`/${id}.html`)
		html = html.replace(/\n/g, '')

		const $ = cheerio.load(html)

		const textEl = $('#entrytext')

		const excerpt = truncate(textEl.text().trim(), { length: 100 })

		textEl.find(`*:not(${Object.keys(this.supportTagsWithAttrs).join(', ')})`).remove()

		for (const [selector, attrList] of Object.entries(this.supportTagsWithAttrs)) {
			const foundElements = textEl.find(selector).toArray()
			for (const element of foundElements) {
				element.attribs = pick(element.attribs, attrList)
			}
		}

		const images = textEl.find('img').toArray()

		const [firstImage] = images

		const previewImageUrl = firstImage ? $(firstImage).attr('src') : undefined

		await Promise.all(
			images.map(async image => {
				const img = $(image)
				const src = img.attr('src')

				const dimensions = src ? await this.getImageSize(src) : null

				if (dimensions && src) {
					image.attribs = {}
					img.attr('src', src)
					img.attr('width', String(dimensions.width))
					img.attr('height', String(dimensions.height))
				} else {
					img.remove()
				}
			}),
		)

		for (const link of textEl.find('a').toArray()) {
			const linkEl = $(link)
			const href = linkEl.attr('href')!
			link.attribs = {}
			linkEl.attr('href', href)
		}

		const title = $('.j-e-title').text().trim()
		const createdAt = this.parseDate($('time[itemprop="dateCreated"]').first())

		const tags = $('.j-e-tags-item')
			.toArray()
			.map(el => $(el).text().trim())

		return {
			id,
			excerpt,
			previewImageUrl: previewImageUrl || null,
			title,
			tags,
			text: textEl.html() || '',
			createdAt: createdAt ? createdAt.toISOString() : null,
		}
	}
}

export const varlamovClient = new VarlamovClient({ endpoint: 'https://varlamov.ru' })
