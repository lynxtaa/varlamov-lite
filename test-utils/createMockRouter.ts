import { NextRouter } from 'next/router'

export function createMockRouter(options?: Partial<NextRouter>): NextRouter {
	return {
		basePath: '/',
		pathname: '/',
		route: '/',
		query: {},
		asPath: '/',
		push: vi.fn(() => Promise.resolve(true)),
		replace: vi.fn(() => Promise.resolve(true)),
		reload: vi.fn(),
		prefetch: vi.fn(() => Promise.resolve()),
		back: vi.fn(),
		forward: vi.fn(),
		beforePopState: vi.fn(),
		isFallback: false,
		isLocaleDomain: true,
		events: {
			on: vi.fn(),
			off: vi.fn(),
			emit: vi.fn(),
		},
		isReady: true,
		isPreview: false,
		...options,
	}
}
