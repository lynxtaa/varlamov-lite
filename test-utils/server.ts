import { setupServer } from 'msw/node'

const server = setupServer()

export { rest } from 'msw'
export { server }
