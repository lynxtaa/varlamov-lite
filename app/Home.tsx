'use client'

import { useInfiniteQuery, useQueryClient } from '@tanstack/react-query'
import { useEffect, useRef } from 'react'

import ArticleListItem from '../components/ArticleListItem'
import Page from '../components/layouts/Page'
import { useIsOnScreen } from '../hooks/useIsOnScreen'
import { Article } from '../lib/varlamovClient'

async function fetchArticles(lastArticleId: number | undefined) {
	const qs =
		lastArticleId !== undefined
			? new URLSearchParams({ lastArticle: String(lastArticleId) })
			: null
	const response = await fetch(`/api/articles${qs ? `?${qs.toString()}` : ''}`)
	if (!response.ok) {
		throw new Error(`Error requesting ${response.url}: ${response.status}`)
	}
	const articles = (await response.json()) as Article[]
	return articles
}

const staleTime = 30 * 60 * 1000
const cacheTime = 30 * 60 * 1000

export default function Home({ initialData }: { initialData: Article[] }) {
	const queryClient = useQueryClient()

	const { data, error, fetchNextPage, hasNextPage, isFetchingNextPage } =
		useInfiniteQuery<Article[], Error>(
			['articles'],
			({ pageParam }: { pageParam?: number }) =>
				queryClient.fetchQuery(
					['next-articles', pageParam],
					() => fetchArticles(pageParam),
					{ staleTime, cacheTime },
				),
			{
				getNextPageParam: lastPage => lastPage[lastPage.length - 1]?.id,
				staleTime,
				cacheTime,
				initialData: {
					pages: [initialData],
					pageParams: [undefined],
				},
			},
		)

	const buttonRef = useRef<HTMLButtonElement>(null)

	const isButtonVisible = useIsOnScreen(buttonRef.current)

	const articles = data?.pages.flat()

	const lastArticleId = articles?.[articles.length - 1]?.id

	useEffect(() => {
		if (
			lastArticleId !== undefined &&
			!isFetchingNextPage &&
			isButtonVisible &&
			hasNextPage === true
		) {
			void queryClient.prefetchQuery(
				['next-articles', lastArticleId],
				() => fetchArticles(lastArticleId),
				{ staleTime, cacheTime },
			)
		}
	}, [hasNextPage, isButtonVisible, isFetchingNextPage, lastArticleId, queryClient])

	return (
		<Page className="max-w-xl" isHome>
			{articles ? (
				articles.length > 0 ? (
					articles.map(article => <ArticleListItem key={article.id} {...article} />)
				) : (
					<header className="mb-8">
						<h2 className="mb-1 text-xl">Ничего не найдено</h2>
					</header>
				)
			) : error ? (
				<span>{error.message}</span>
			) : null}
			{hasNextPage === true && (
				<button
					className="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline"
					onClick={() => {
						void fetchNextPage()
					}}
					disabled={isFetchingNextPage}
					ref={buttonRef}
				>
					Показать больше {isFetchingNextPage && '...'}
				</button>
			)}
		</Page>
	)
}
