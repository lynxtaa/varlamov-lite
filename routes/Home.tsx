import { useInfiniteQuery } from 'react-query'

import ArticleListItem from '../components/ArticleListItem'
import Page from '../components/layouts/Page'
import { Article } from '../lib/varlamovClient'

async function fetchArticles({ pageParam }: { pageParam?: number }) {
	const qs = pageParam ? new URLSearchParams({ lastArticle: String(pageParam) }) : null
	const response = await fetch(`/api/articles${qs ? `?${qs}` : ''}`)
	if (!response.ok) {
		throw new Error(`Error requesting ${response.url}: ${response.status}`)
	}
	const articles = (await response.json()) as Article[]
	return articles
}

export default function Home() {
	const { data, error, fetchNextPage, hasNextPage, isFetchingNextPage } =
		useInfiniteQuery<Article[], Error>('articles', fetchArticles, {
			getNextPageParam: lastPage => lastPage[lastPage.length - 1]?.id,
			staleTime: 30 * 60 * 1000,
			cacheTime: 30 * 60 * 1000,
		})

	const articles = data && data.pages.flat()

	return (
		<Page className="max-w-xl">
			{articles ? (
				articles.length > 0 ? (
					articles.map(article => <ArticleListItem key={article.id} {...article} />)
				) : (
					<header className="mb-8">
						<h2 className="mb-1 text-xl">Ничего не найдено</h2>
					</header>
				)
			) : error ? (
				<span>{error.message}</span>
			) : null}
			{hasNextPage && (
				<button
					className="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline"
					onClick={() => fetchNextPage()}
					disabled={isFetchingNextPage}
				>
					Показать больше {isFetchingNextPage && '...'}
				</button>
			)}
		</Page>
	)
}
