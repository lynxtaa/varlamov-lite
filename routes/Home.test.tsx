import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { subDays, subHours, subMinutes } from 'date-fns'
import { advanceTo } from 'jest-date-mock'
import { DefaultRequestBody } from 'msw'
import { NextRouter, useRouter } from 'next/router'

import { createMockRouter } from '../jest/createMockRouter'
import { renderWithProviders } from '../jest/renderWithProviders'
import { server, rest } from '../jest/server'
import { Article } from '../lib/varlamovClient'
import Home from '../pages/index'

const useRouterMock = useRouter as jest.Mock<NextRouter>

let mockRouter

beforeEach(() => {
	mockRouter = createMockRouter()
	useRouterMock.mockReturnValue(mockRouter)
})

it('shows blog posts', async () => {
	const now = new Date('2020-03-20')
	advanceTo(now)

	server.use(
		rest.get<DefaultRequestBody, Article[]>('/api/articles', (req, res, ctx) => {
			const articles: Article[] = [
				{
					id: 1,
					uri: '1.html',
					title: 'Первая новость',
					published_at: subMinutes(now, 2).toISOString(),
				},
				{
					id: 2,
					uri: '2.html',
					title: 'Вторая новость',
					published_at: subHours(now, 2).toISOString(),
				},
				{
					id: 3,
					uri: '3.html',
					title: 'Третья новость',
					published_at: subDays(now, 2).toISOString(),
				},
			]

			return res(ctx.json(articles))
		}),
	)

	const container = renderWithProviders(<Home />)

	await screen.findByText('Первая новость')

	expect(container).toMatchSnapshot()
})

it('runs search with entered query', async () => {
	server.use(
		rest.get<DefaultRequestBody, Article[]>('/api/articles', (req, res, ctx) =>
			res(ctx.json([])),
		),
	)

	renderWithProviders(<Home />)

	await screen.findByText('Ничего не найдено')

	userEvent.click(screen.getByRole('button', { name: 'Поиск' }))
	userEvent.type(screen.getByRole('textbox', { name: 'Поиск' }), 'moscow')
	userEvent.click(screen.getByRole('button', { name: 'Отправить' }))

	await waitFor(() => {
		expect(mockRouter.push).toHaveBeenCalledWith('/search/moscow')
	})
})
