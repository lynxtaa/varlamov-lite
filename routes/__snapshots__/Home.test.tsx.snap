// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows blog posts 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="my-0 mx-auto p-4 max-w-xl"
      >
        <header
          class="flex flex-col mt-4 mb-8"
        >
          <div
            class="flex items-center"
          >
            <h1
              class="text-2xl mr-3 font-bold"
            >
              Блог Ильи Варламова
            </h1>
            <button
              class="ml-auto"
              title="Поиск"
              type="button"
            >
              <div
                class="w-7 h-7"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  style="width: 100%; height: 100%;"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="11"
                    cy="11"
                    r="8"
                  />
                  <line
                    x1="21"
                    x2="16.65"
                    y1="21"
                    y2="16.65"
                  />
                </svg>
              </div>
            </button>
            <button
              class="ml-4"
              title="Переключить тему"
              type="button"
            >
              <div
                class="w-7 h-7"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  style="width: 100%; height: 100%;"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
                  />
                </svg>
              </div>
            </button>
          </div>
        </header>
        <header
          class="mb-8"
        >
          <h2
            class="mb-1 text-xl"
          >
            <a
              class="hover:underline"
              href="/blog/1.html"
            >
              Первая новость
            </a>
          </h2>
          <time
            class="opacity-70"
          >
            сегодня в 2:58
          </time>
        </header>
        <header
          class="mb-8"
        >
          <h2
            class="mb-1 text-xl"
          >
            <a
              class="hover:underline"
              href="/blog/2.html"
            >
              Вторая новость
            </a>
          </h2>
          <time
            class="opacity-70"
          >
            сегодня в 1:00
          </time>
        </header>
        <header
          class="mb-8"
        >
          <h2
            class="mb-1 text-xl"
          >
            <a
              class="hover:underline"
              href="/blog/3.html"
            >
              Третья новость
            </a>
          </h2>
          <time
            class="opacity-70"
          >
            18 марта
          </time>
        </header>
        <button
          class="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline"
        >
          Показать больше 
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="my-0 mx-auto p-4 max-w-xl"
    >
      <header
        class="flex flex-col mt-4 mb-8"
      >
        <div
          class="flex items-center"
        >
          <h1
            class="text-2xl mr-3 font-bold"
          >
            Блог Ильи Варламова
          </h1>
          <button
            class="ml-auto"
            title="Поиск"
            type="button"
          >
            <div
              class="w-7 h-7"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="width: 100%; height: 100%;"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
                <line
                  x1="21"
                  x2="16.65"
                  y1="21"
                  y2="16.65"
                />
              </svg>
            </div>
          </button>
          <button
            class="ml-4"
            title="Переключить тему"
            type="button"
          >
            <div
              class="w-7 h-7"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="width: 100%; height: 100%;"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
                />
              </svg>
            </div>
          </button>
        </div>
      </header>
      <header
        class="mb-8"
      >
        <h2
          class="mb-1 text-xl"
        >
          <a
            class="hover:underline"
            href="/blog/1.html"
          >
            Первая новость
          </a>
        </h2>
        <time
          class="opacity-70"
        >
          сегодня в 2:58
        </time>
      </header>
      <header
        class="mb-8"
      >
        <h2
          class="mb-1 text-xl"
        >
          <a
            class="hover:underline"
            href="/blog/2.html"
          >
            Вторая новость
          </a>
        </h2>
        <time
          class="opacity-70"
        >
          сегодня в 1:00
        </time>
      </header>
      <header
        class="mb-8"
      >
        <h2
          class="mb-1 text-xl"
        >
          <a
            class="hover:underline"
            href="/blog/3.html"
          >
            Третья новость
          </a>
        </h2>
        <time
          class="opacity-70"
        >
          18 марта
        </time>
      </header>
      <button
        class="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline"
      >
        Показать больше 
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
